ENH: Implement initial kwarg for ufunc.add.reduce

hameerabbasi opened the pull request:
As discussed in #5032: Adds an `initial` kwarg for `ufunc.reduce`, plus changes in `min`, `max`, `sum`, `prod`.

This acts as the first item in a reduction. It's useful for empty reductions e.g. `np.min([], initial=np.inf)` reduces to `inf`. It's also useful for starting the reduction with a different value e.g. `np.sum([10], initial=5)` reduces to `15`. (`5` used as starting value rather than the identity `0`).

Explicitly passing the initial value as `None` causes empty reductions to err.

Explicitly passing `np._NoValue` also keeps the default behaviour.

`min`, `max`, `sum` and `prod` only pass on the `initial` kwarg.

`PyUFunc_Reduce` is modified to accept the `initial` kwarg. It then replaces the identity by `initial`.

Changes are made to `normalize_reduce_args` so that it accepts and parses `initial` correctly.

Documentation has been added for all the changes.

Fixes #5032

hameerabbasi commented:
@eric-wieser Looking forward to your feedback. 😄

hameerabbasi commented:
Hmm. I have no idea why the CircleCI build failed, I didn't touch Cython. :/

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -2844,7 +2843,9 @@ PyUFunc_Reduce(PyUFuncObject *ufunc, PyArrayObject *arr, PyArrayObject *out,
-    identity = _get_identity(ufunc, &reorderable);
+    if (identity == NULL || identity == Py_None) {
+        identity = _get_identity(ufunc, &reorderable);
+    }
```
You need an `else { Py_INCREF(identity); }`, so that the decref at the end is safe.

hameerabbasi replied:
I don't think so, it only checks for `None` and `NULL`. The object is getting passed through.

`identity` is gotten from the `PyArg_ParseTupleAndKeywords`, does this do the `Py_INCREF` call automatically or not?

eric-wieser replied:
And the other half of the puzzle is that `_get_identity` _does_ increment the reference count - so you want to make them equivalent so the cleanup doesn't need special casing.

eric-wieser commented:
It looks like  you built this patch on an out-of-date master from before we had CircleCI working. If you rebase, the CircleCI error will go away.

eric-wieser commented:
This is failing because we reparse the arguments for forwarding to __array_ufunc__ - but I don't remember where.

ahaldane commented:
Somewhat tangent comment: How would this fit in with the dream of turning `reduce` into a gufunc (related to discussions in #8528 #8819)? 

Currently gufuncs have a fixed set of common keyword arguments, `axis`, `dtype`, `casting`, `extobj`, `where` and so on. Should individual gufuncs be allowed to have additional args besides the default ones, as `identity` would be?

Or does the fact that reduce will have an extra `identity` argument mean that it cannot be implemented as a gufunc?

ahaldane commented:
Maybe the way to think of it is that a `reduce_ufunc(a, out=b)` gufunc treats its "out" parameter also as an input. So if you supply an `out`, that initial value gets read and added on to. 

So maybe ultimately, we would have both a reduce-gufunc, and a reduce method. The reduce method's job is to properly set up the out parameter for the reduce-gufunc. The method would have an "identity" argument, while the ufunc itself would read that value from the out arg.

Sorry for the distraction, we can think about this later outside this PR.

charris commented:
The explanation of these changes could be much more complete. And please explain it for *all* the functions changed.

eric-wieser commented:
So, there are two things that worry me here:

* The interface is different to the builtin `max(..., default=0)`, even though it's _roughly_ the same
* `ufunc.reduce(identity=...)` only really makes sense for `maximum`. `minimum`, `fmax`, and `fmin`, where choosing a different domain lets you choose a different identity. For other ufuncs, it produces crazy results (I think `np.add.reduce(np.ones(2,2), identity=10)` will give a surprising result, but I haven't checked).

hameerabbasi commented:
> The explanation of these changes could be much more complete. And please explain it for all the functions changed.

I'll add more complete docs with examples, if that's what you mean, incorporating @eric-wieser's excellent use cases and insights. :-)

> The interface is different to the builtin max(..., default=0), even though it's roughly the same.

Yes, I thought of that as well, however, plenty of kwargs here are different (`axis`, `keepdims`, etc.). I'll go ahead and assume that we're not following Python's interface here, however, this can still be changed.

However, if it WAS to be changed, I'd say it's more similar to [`functools.reduce`](https://docs.python.org/3/library/functools.html#functools.reduce)'s `initializer` argument, which is much more instructive. This is also in line with your second worry:

> For other ufuncs, it produces crazy results (I think np.add.reduce(np.ones(2,2), identity=10) will give a surprising result, but I haven't checked).

```python
>>> np.add.reduce(np.ones((2, 2)), identity=10)
array([12., 12.])
```

The name `initializer` instead of `identity` is much more instructive and takes care of this worry.

So `initializer` probably makes more sense here. We could add a warning in the docs that non-default initializers could lead to unexpected results.

eric-wieser commented:
> will give a surprising result, but I haven't checked

There's a more surprising result than that one to be found, I think. Maybe `np.add.reduce(np.ones((2, 2, 2)), axis=(0, 1), identity=10)`?

hameerabbasi commented:
It gives a quite reasonable result, actually.
```python
>>> np.add.reduce(np.ones((2, 2, 2)), axis=(0, 1), identity=10)
array([14., 14.])
```

eric-wieser commented:
> The name initializer instead of identity is much more instructive and takes care of this worry.

That's maybe reasonable. Perhaps `ufunc.reduce` should take an `initializer` argument, but the `np.min` shorthand can alias it to `default`?

We'll definitely need to run this by the list before this gets merged, but

eric-wieser commented:
Non-adjacent axes? `(0, 2)`?

hameerabbasi commented:
> Non-adjacent axes? `(0, 2)`?

Still works beautifully.

> That's maybe reasonable. Perhaps ufunc.reduce should take an initializer argument, but the np.min shorthand can alias it to default?

The slight problem I see with this is something like the following:

```python
>>> np.minimum.reduce([0.0], identity=-np.inf)
-inf
```

As you can see, it gives a value not in the original set. I guess we could call it `default` but then there are two issues with this:

- The one I mentioned above;
- We're deviating from `ufunc.reduce` norms, which is a bad thing. It doesn't keep uniformity.

eric-wieser commented:
> Still works beautifully.

Huh, I think my memory of the code is incorrect. I thought that multi-axis `reduce` fell back on repeated single-axis reduce, and so was expecting to get 24 as the result. It seems that's not the case, which is great!

> As you can see, it gives a value not in the original set.

I'm now on board with the argument being called `initializer`, which makes that far less confusing.

The issue in that case is that the caller chose to pass in an initializer that was _not_ the minimum of the set of allowed values. That looks perfectly fine to me.

eric-wieser commented:
CI is failing because `core/src/umath/override.c@normalize_reduce_args` is missing this argument.

hameerabbasi commented:
> The issue in that case is that the caller chose to pass in an initializer that was not the minimum of the set of allowed values. That looks perfectly fine to me.

Exactly. We can document a use-case as being reduction of an empty axis.

On that note, I'm going to work and might not be available for the next hours. If you want to fix that on my branch, feel free.

hameerabbasi commented:
On another note, if we're renaming this to `initializer`, it would make sense to add it to `accumulate` and `reduceat`. Is that best done in another PR or this one?

eric-wieser commented:
`reduceat` is best untouched - there's an issue about writing a replacement somewhere.

Accumulate would be a better fit for another PR, especially since the use case is less strong.

eric-wieser commented:
> Edit: I'm willing to take on the task of making reduceat use identity.

#834 should get you started on that rabbit hole. Lets try and wrap up this PR first though!

eric-wieser commented on numpy/add_newdocs.py:
```diff
@@ -5874,7 +5880,19 @@ def luf(lamdaexpr, *args, **kwargs):
+           
+    You can use the initializer argument to initialize the reduction to a
+    different value. It can be used to calculate empty reductions for
+    ufuncs that don't support it as well.
```
Move this second sentence to just above `np.minimum.reduce` so it's clear what it refers to

hameerabbasi replied:
Done!

eric-wieser commented on numpy/core/fromnumeric.py:
```diff
@@ -1827,6 +1827,11 @@ def sum(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
         exceptions will be raised.
+    initializer : scalar, optional
+        The initializer for this maximum. Can also be used to compute the
+        minimum of an empty slice.
```
`max`? `min`? This is `sum`

I'd argue that this isn't useful enough to include in the public API here, or in any of these functions. It's niche enough that we probably want the user calling `ufunc.reduce` directly

hameerabbasi replied:
Hmm. I went to a lot of effort to put it in. :-( Anyway, I'd argue it should stay for the NaN-skipping functions because there's no alternative there.

hameerabbasi replied:
Additionally, it's useful for specifying the value of an All-NaN slice reduction (although I guess you could just do this):

```python
x_max = np.nanmax(x, ...)
x[np.isnan(x)] = value
```

eric-wieser replied:
Here's what I propose:

* `np.some_ufunc.reduce(..., initializer=...)`
* `np.sum(...)`
* `np.prod(...)`
* `np.min(..., default=...)` - forward to initializer, match the py3 API
* `np.nanmin(..., default=...)` as above

eric-wieser replied:
I take it back, `default` and `initializer` mean different things. `max([0], default=10)` is different to `np.max([0], initializer=10)`

hameerabbasi replied:
Exactly, as in the discussion above. But to keep uniformity, we should probably either keep it or remove it everywhere. I prefer users should not have to remember/recall which reductions had this and which didn't, they should just instinctively go for either the reduction or `ufunc.reduce`.

I guess to cater for the `nanmax`, etc. case, we could ask users to reshape, add an extra slice, then undo, but that would be excessive.

eric-wieser replied:
nanmax absolutely should have an `initializer` argument, since it needs it to deal with all-nan slices of object dtypes. I'm just not sure they're useful on `sum` and `prod`, but maybe leave them in for now, and consult the list.

Either way, my initial remark here is pointing out that this docstring is copied-and-pasted incorrectly

hameerabbasi replied:
That's fixed. I see you noticed it was a copypasta.

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -3275,7 +3284,7 @@ PyUFunc_Reduceat(PyUFuncObject *ufunc, PyArrayObject *arr, PyArrayObject *ind,
     npy_uint32 op_flags[3];
     int i, idim, ndim, otype_final;
-    int need_outer_iterator;
+    int need_outer_iterator = 0;
```
Why this change?

hameerabbasi replied:
There was a warning in GCC, I wanted to get rid of it.

hameerabbasi replied:
Are the warnings/errors because of this change? Some C++ magic going on? Ideally, adding an intitializer shouldn't change anything.

eric-wieser replied:
The warnings are showing up because there's a failure elsewhere - you're probably just not scrolling down far enough to see the error. I'd probably revert this change, and leave it for a future warning cleanup.

hameerabbasi commented:
Seems like this needs a rebase. I'll push as soon as the few discussions reach a consensus. Locally, all tests pass for me.

hameerabbasi commented:
Hmm. These last failures have nothing at all to do with my code. :/

hameerabbasi commented:
Okay, this time, I really; truly give up. I now believe these failures had nothing at all to do with my code.

eric-wieser commented:
There are real errors here that relatedto your code: https://ci.appveyor.com/project/charris/numpy/build/1.0.9149/job/kk5xfpmmnfo0gdd1/tests

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -2864,7 +2864,16 @@ PyUFunc_Reduce(PyUFuncObject *ufunc, PyArrayObject *arr, PyArrayObject *out,
+        /* Since initializer is passed through from the argument parsing,
+           we must increment the reference counter. */
+        identity = initializer;
+        Py_INCREF(initializer);
```
This branch doesn't set `reorderable`, which is what is causing (intermittent) bugs.

hameerabbasi commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -2864,7 +2864,16 @@ PyUFunc_Reduce(PyUFuncObject *ufunc, PyArrayObject *arr, PyArrayObject *out,
 
     /* Get the identity */
-    identity = _get_identity(ufunc, &reorderable);
+    if (initializer == NULL || initializer == Py_None) {
```
That's a toughie. Thing is, we pass through `None` for missing values on many counts. If there was a way to default to `np._NoValue` which would go to `NULL`, I'd be for it.

eric-wieser replied:
Maybe something to defer to the mailing list. Let's get the `reorderable` thing passing first

eric-wieser commented on numpy/add_newdocs.py:
```diff
@@ -5787,7 +5787,7 @@ def luf(lamdaexpr, *args, **kwargs):
 add_newdoc('numpy.core', 'ufunc', ('reduce',
     """
-    reduce(a, axis=0, dtype=None, out=None, keepdims=False)
+    reduce(a, axis=0, dtype=None, out=None, keepdims=False, initialization=None)
```
Should say `initializer`

hameerabbasi replied:
Done!

eric-wieser commented on numpy/core/_methods.py:
```diff
@@ -22,17 +22,17 @@
 # small reductions
-def _amax(a, axis=None, out=None, keepdims=False):
-    return umr_maximum(a, axis, None, out, keepdims)
+def _amax(a, axis=None, out=None, keepdims=False, initializer=None):
```
Oh, the easy fix here is just to write `initializer=um.maximum.identity`, and avoid having `None` mean default

hameerabbasi replied:
Would that potentially break duck arrays? Passing `initializer` when it's not supposed to be passed?

hameerabbasi replied:
It might already be handling this down below.

```
if (initializer != Py_None && PyArray_ISOBJECT(arr) && PyArray_SIZE(arr) != 0) {
        Py_DECREF(initializer);
        initializer = Py_None;
        Py_INCREF(initializer);
    }
```

eric-wieser replied:
It would affect `__array_ufunc__`, I suppose.

I think there might be a doc section on that that will need updating too.

hameerabbasi replied:
Oh, wait, it wouldn't. I retract my objection, those are handled in `fromnumeric.py`

hameerabbasi commented:
> I think you want:
>```
>...
>```

This (combined with replacing `identity` with `initializer` everywhere below) causes datatype failures. One because it can't add `int` and `str` and one because something that should be `bool` isn't.

eric-wieser commented:
Actually I think `initializer=None` is not supported at a bunch of levels anyway, so I don't think changing the default arguments really made any difference. Perhaps `None` was better after all as a "don't use this" flag

hameerabbasi commented:
This might be slightly off-topic, but:

- What IDE do you recommend for Python API C++.
- Can I see the built devdocs somewhere?

hameerabbasi commented:
> Actually I think initializer=None is not supported at a bunch of levels anyway,

I'm not so sure. I used `np._NoValue` as the default rather than `None` everywhere, and only added it to kwargs if it wasn't `np._NoValue`. I don't exactly know how to test this (we'll need #8955 merged to know for sure).

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -2865,35 +2865,44 @@ PyUFunc_Reduce(PyUFuncObject *ufunc, PyArrayObject *arr, PyArrayObject *out,
 
     /* Get the identity */
     identity = _get_identity(ufunc, &reorderable);
```
You need to check if this returns `NULL` like it did before, because that means that `reorderable` wasn't computed either.

hameerabbasi replied:
That's already done below; I think. I changed `identity` to `initializer` everywhere below, so all checks are now on `initializer`.

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -2865,35 +2865,44 @@ PyUFunc_Reduce(PyUFuncObject *ufunc, PyArrayObject *arr, PyArrayObject *out,
+
+    if (initializer == NULL) {
         return NULL;
     }
```
This needs to happen _before_ you do the swap, otherwise `identity` might have been `NULL` and `DECREF`'d, which would segfault

hameerabbasi replied:
I've done it (and no tests fail locally), but just so I understand: Does `NULL` mean that something went wrong and the reduction failed?

eric-wieser replied:
If a C function returns `NULL`, that usually means it set a python exception. If that happens, the right thing to do is usually to clean up, and then return `NULL` all the way back to the calling function.

Have a look at `_get_identity` to see how it can return `NULL`.

hameerabbasi commented:
Oh. That fixed the CI as well. Thanks!

eric-wieser commented:
Something to keep on your radar - this argument can be used to fix some cases in `linalg.norm` that invoke `max`

eric-wieser commented on numpy/add_newdocs.py:
```diff
@@ -5843,6 +5843,12 @@ def luf(lamdaexpr, *args, **kwargs):
 
         .. versionadded:: 1.7.0
+    initializer : scalar, optional
+        The initializer before starting the reduction. The reduction starts
```
How about

> The value to which the result is initialized before starting the reduction. If not specified, the identity for this ufunc is used. For ufuncs with no identity like `maximum`, a reduction cannot be performed over an empty axis unless this argument is provided.

hameerabbasi replied:
> If None

Would "if not supplied" be better for object arrays in the future?

eric-wieser replied:
My comment above describes what actually happens if `None` is passed, I think - it may not be what we want, but it's what we have right now.

I think the eventual fix here is to provide `np.ufunc.NoIdentity` as a sentinel that we use in place of the current `None`

hameerabbasi replied:
I think that should technically go to #8955.

eric-wieser replied:
One final remark for this docstring
```
If arr.dtype is object, then the initializer is _only_ used when the reduction is empty
```

This is perhaps not ideal, but importantly it _is_ useful enough to fix `linalg.norm`

hameerabbasi replied:
I expected this to not work for empty `object` arrays at all, but it seems you're right! I've updated the docstring.

```python
>>> class A(object):
...     pass
... 
>>> np.sum(np.array([], dtype=object), initializer=A())
<A object at 0x1023abf28>
```

eric-wieser replied:
I was expecting exactly the same thing, and was half-way through writing about how that was going to be really annoying before realizing we can sort-of get away with it

hameerabbasi commented:
Which cases specifically? If you open an issue with a specific case I'll look into it. I can assume you're talking about either the L0 (min) or the Linf (max) norm, but not sure what specific cases you mean.

eric-wieser commented on numpy/core/fromnumeric.py:
```diff
@@ -2343,11 +2362,21 @@ def amax(a, axis=None, out=None, keepdims=np._NoValue):
+    >>> np.max([], initializer=-np.inf)
+    -inf
+    >>> np.max([5], initializer=6)
+    6
```
Comparison to the python 3 `max(default=...)` would be good here. Same for `min`

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -2863,37 +2863,47 @@ PyUFunc_Reduce(PyUFuncObject *ufunc, PyArrayObject *arr, PyArrayObject *out,
+    }
+    else {
+        Py_DECREF(identity);
+        Py_INCREF(initializer);  /* match the reference count in the of above */
```
`of` should read `if`

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -2863,37 +2863,47 @@ PyUFunc_Reduce(PyUFuncObject *ufunc, PyArrayObject *arr, PyArrayObject *out,
     }
 
-    /* Get the identity */
+    /* Check if identity makes sense */
```
Not really accurate. `_get_identity` only fails if something really bad has happened, either:

* Someone has created an invalid ufunc
* `PyInt_FromLong(1)` failed

I'd prefer to leave the original comment

eric-wieser commented:
@hameerabbasi: I think both are well-defined? Linf = 0, L0 = inf seem pretty reasonable to me. #10655 tries to fix this in a hacky way, but an approach relying on this PR would be much cleaner.

Just some documentation remarks left, then I'll try to compose a message to the list

eric-wieser commented:
I'm expecting `np.sum([], initializer=None)` to give an error - I think that's ok, but it would be good to test it.

> We run into the same issue there (although what you're saying is more reasonable than here)

I think that argument isn't valid, because the norms are defined over the real numbers, so the set of allowable values is known. But it's a good point to raise, and one I hadn't thought of. If you want to continue that discussion, create a new issue about it - I can't seem to find the old one.

hameerabbasi commented:
> I'm expecting np.sum([], initializer=None) to give an error - I think that's ok, but it would be good to test it.

It does.

```python
>>> np.sum([], initializer=None)
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "/Users/hameerabbasi/PycharmProjects/numpy/numpy/core/fromnumeric.py", line 1905, in sum
    out=out, **kwargs)
  File "/Users/hameerabbasi/PycharmProjects/numpy/numpy/core/_methods.py", line 35, in _sum
    return umr_sum(a, axis, dtype, out, keepdims, initializer)
ValueError: zero-size array to reduction operation add which has no identity
```

I've documented it:

```
If ``None`` is given to a reduction for an empty array, it will raise an error.
```

hameerabbasi commented:
Let me know when no further changes are desired and I'll commit. Don't want to hog the CI too much.

eric-wieser commented:
I wouldn't worry too much about the CI - timezones tend to even things out.

eric-wieser commented on numpy/add_newdocs.py:
```diff
@@ -5844,9 +5844,12 @@ def luf(lamdaexpr, *args, **kwargs):
+        Defaults to the `~numpy.ufunc.identity` of the ufunc.
+        If ``None`` is given, then the first element of the reduction is used,
+        and an error is thrown if the reduction is empty. If ``None`` is given to a
+        reduction for a empty axes, it will raise an error. If ``a.dtype`` is ``object``,
```
Aren't these two sentences the same?

>  If ``None`` is given, then the first element of the reduction is used, and an error is thrown if the reduction is empty.

and

> If ``None`` is given to a reduction for a empty axes, it will raise an error.

hameerabbasi replied:
Fixed!

hameerabbasi commented:
> norms are defined over the real numbers

1. They're defined over an arbitrary set of numbers/vectors/complexes/Cayley-Dickson numbers etc.
2. The input set of Linf might actually reduce to a set of reals that's _not_ bounded below by zero. It might be bounded below by some other quantity.
3. L0 doesn't have this issue, it's a sum rather than a min (as I incorrectly assumed).

eric-wieser commented on numpy/lib/nanfunctions.py:
```diff
@@ -395,11 +415,12 @@ def nanmax(a, axis=None, out=None, keepdims=np._NoValue):
+            mask = np.all(mask, axis=axis, **kwargs)
+            if np.any(mask):
+                res = _copyto(res, np.nan, mask)
+                warnings.warn("All-NaN axis encountered", RuntimeWarning, stacklevel=2)
```
What if `initializer` is NaN?

hameerabbasi replied:
Hmm. I was unsure what to do there, actually. On one hand, it may mean that the user WANTS `nan` values in the output. On the other, it would give `nan`s.

eric-wieser replied:
Fine as is for now I think.

eric-wieser replied:
Works for me

hameerabbasi replied:
I fixed the potential issue. Wasn't too hard. :-)

eric-wieser commented on numpy/core/tests/test_ufunc.py:
```diff
@@ -941,6 +946,27 @@ def test_identityless_reduction_noncontig_unaligned(self):
+    def test_initializer_reduction(self):
+        # np.minimum.reduce is an identityless reduction
+
+        # For cases like np.maximum(np.abs(...), identity=0)
```
I think this should say `np.max(np.asb(...), initializer=0)`

hameerabbasi replied:
Done!

eric-wieser commented on numpy/core/fromnumeric.py:
```diff
@@ -2529,10 +2641,16 @@ def prod(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
     >>> np.prod(x).dtype == int
     True
 
+    You can also initialize the product by a different value.
```
I'd argue it's "initialize with", not "initialize by"

hameerabbasi replied:
Done!

eric-wieser commented on numpy/core/fromnumeric.py:
```diff
@@ -1895,6 +1905,56 @@ def sum(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
+    return arr.any(axis=axis, out=out, **kwargs)
+
+
+def alltrue(a, axis=None, out=None, keepdims=np._NoValue):
```
Bad rebase here

eric-wieser reviewed (dismissed):
Something went wrong with your rebase

eric-wieser commented on numpy/core/fromnumeric.py:
```diff
@@ -2230,6 +2290,13 @@ def amax(a, axis=None, out=None, keepdims=np._NoValue):
 
+    initializer : scalar, optional
+        The initializer for this maximum. Can also be used to compute the
+        maximum of an empty slice.
```
Perhaps we should say `` See `~numpy.ufunc.reduce` for details`` or something to indicate that this is a simplified summary.

hameerabbasi replied:
Done!

eric-wieser commented on numpy/core/fromnumeric.py:
```diff
@@ -1895,6 +1905,22 @@ def sum(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
                          out=out, **kwargs)
 
 
+def product(a, axis=None, dtype=None, out=None, keepdims=np._NoValue, initializer=np._NoValue):
```
This function should be removed - it's now defined with `*args, **kwargs` at the bottom of the file.

eric-wieser commented on numpy/lib/nanfunctions.py:
```diff
@@ -644,7 +678,12 @@ def nanprod(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
     a, mask = _replace_nan(a, 1)
-    return np.prod(a, axis=axis, dtype=dtype, out=out, keepdims=keepdims)
+
+    if np.isnan(initializer):
```
`isnan` doesn't work for `np._NoValue`. I'm not sure it's clear that nanprod` should treat initializer specially, especially since you don't document this. Keep it simple for now, without this check, I'd say.

eric-wieser commented on numpy/lib/nanfunctions.py:
```diff
@@ -877,7 +916,7 @@ def _nanmedian1d(arr1d, overwrite_input=False):
     """
     arr1d, overwrite_input = _remove_nan_1d(arr1d,
-        overwrite_input=overwrite_input)
+                                            overwrite_input=overwrite_input)
```
nit: I preferred the previous indentation because it's more robust to variable renames.

hameerabbasi replied:
My IDE's auto-format did it. :/

eric-wieser commented on numpy/lib/nanfunctions.py:
```diff
@@ -644,7 +678,12 @@ def nanprod(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
     a, mask = _replace_nan(a, 1)
-    return np.prod(a, axis=axis, dtype=dtype, out=out, keepdims=keepdims)
+
+    if initializer != np._NoValue:
```
Should use `is not` with `np._NoValue`

eric-wieser commented on numpy/lib/nanfunctions.py:
```diff
@@ -616,6 +647,9 @@ def nanprod(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
         broadcast correctly against the original `arr`.
+    initializer : scalar, optional
+        The value to initialize the product to. See `~numpy.ufunc.reduce`
+        for details.
```
Missing `.. versionadded`

eric-wieser commented on numpy/lib/nanfunctions.py:
```diff
@@ -327,6 +337,12 @@ def nanmax(a, axis=None, out=None, keepdims=np._NoValue):
+    initializer : scalar, optional
+        The value to initialize the maximum to. This can also be used to compute
+        the maximum of empty slices. If the initializer is given, no warning is
+        issued. See `~numpy.ufunc.reduce` for details.
```
is issued **for an all-NaN slice**.

mhvk commented on numpy/core/fromnumeric.py:
```diff
@@ -1827,6 +1827,10 @@ def sum(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
         sub-class' method does not implement `keepdims` any
         exceptions will be raised.
+    initializer : scalar, optional
+        The initializer for this sum. See `~numpy.ufunc.reduce` for details.
```
Repeating the term is not helpful for those who do not understand what it does. Maybe
"Starting value for the sum. See..."

mhvk commented on numpy/core/fromnumeric.py:
```diff
@@ -2230,6 +2239,13 @@ def amax(a, axis=None, out=None, keepdims=np._NoValue):
         exceptions will be raised.
 
+    initializer : scalar, optional
+        The initializer for this maximum. Can also be used to compute the
```
See above. Maybe "Minimum value for the maximum. Needs to be present to allow calculation on an empty slice."

mhvk commented on numpy/core/fromnumeric.py:
```diff
@@ -2282,11 +2298,28 @@ def amax(a, axis=None, out=None, keepdims=np._NoValue):
+
+    >>> np.max([], initializer=-np.inf)
+    -inf
+    >>> np.max([50], initializer=25)
```
Maybe use something a bit more likely here, which shows that the initializer is just included in the comparison. E.g.,
```
np.max([[-50], [10]], axis=-1, initializer=0)
[[0],
 [10]]
```
(guessing at formatting)

mhvk commented on numpy/core/fromnumeric.py:
```diff
@@ -2282,11 +2298,28 @@ def amax(a, axis=None, out=None, keepdims=np._NoValue):
+    >>> np.max([50], initializer=25)
+    50
+
+    Notice that this isn't the same as Python's ``default`` argument.
```
Be explicit in what is different: "Notice that the initializer is used as one of the elements for which the maximum is determined, unlike for the ``default`` argument Python's `max` function, which is only used for empty iterables."

mhvk commented on numpy/core/fromnumeric.py:
```diff
@@ -1874,10 +1878,16 @@ def sum(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
     >>> np.ones(128, dtype=np.int8).sum(dtype=np.int8)
     -128
 
+    You can also use initialize the sum with a different value:
```
Maybe "You can also start the sum with an initial value other than 0:"

mhvk commented on numpy/core/fromnumeric.py:
```diff
@@ -2331,6 +2364,12 @@ def amin(a, axis=None, out=None, keepdims=np._NoValue):
         exceptions will be raised.
 
+    initializer : scalar, optional
+        The initializer for this minimum. Can also be used to compute the
```
See above (here and below)

mhvk commented on numpy/core/fromnumeric.py:
```diff
@@ -2472,6 +2529,10 @@ def prod(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
         sub-class' method does not implement `keepdims` any
         exceptions will be raised.
+    initializer : scalar, optional
+        The initializer for this product. See `~numpy.ufunc.reduce` for details.
```
See comment above for `sum`

mhvk reviewed (commented):
This is a nice addition. Some comments on the docstrings, which I think can be clearer. (will continue the review of the rest, but sending this now since changes are being made...)

mhvk commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -2868,32 +2868,42 @@ PyUFunc_Reduce(PyUFuncObject *ufunc, PyArrayObject *arr, PyArrayObject *out,
+
+    /* Get the initializer */
+    if (initializer == NULL) {
+        initializer = identity;
```
I think the `identity` should only be gotten here, inside the loop. After all, it is not used if an `initializer` is passed on, so one might as well skip getting it.

hameerabbasi replied:
We need `reorderable`, which we get from the same function. I did that at first but was having failures.

mhvk replied:
Ah, that makes sense. That's what one gets for coming to the discussion late...

mhvk commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -2868,32 +2868,42 @@ PyUFunc_Reduce(PyUFuncObject *ufunc, PyArrayObject *arr, PyArrayObject *out,
+        initializer = identity;
+    }
+    else {
+        Py_DECREF(identity);
```
If we do as suggested above, the `Py_DECREF` can obviously be removed.

hameerabbasi replied:
See comment to above.

mhvk commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -2868,32 +2868,42 @@ PyUFunc_Reduce(PyUFuncObject *ufunc, PyArrayObject *arr, PyArrayObject *out,
-        Py_DECREF(identity);
-        identity = Py_None;
-        Py_INCREF(identity);
+    if (initializer != Py_None && PyArray_ISOBJECT(arr) && PyArray_SIZE(arr) != 0) {
```
Why do we do this? I think it is actually rather useful to be able to pass on something else than `None` (but can see that this would not be the case with a built-in `identity`)

hameerabbasi replied:
I'm leaving supporting object arrays to #8955, all the object-specific cases will be handled there.

hameerabbasi replied:
It's the same as the logic before, just on `initializer` instead of identity.

mhvk replied:
OK, fair enough. Could you add a comment in the code still?

hameerabbasi replied:
One is already there. :-)

hameerabbasi commented on numpy/add_newdocs.py:
```diff
@@ -5874,7 +5882,22 @@ def luf(lamdaexpr, *args, **kwargs):
+    array([14., 14.])
+    >>> np.add.reduce([10], initializer=5)
+    15
+    >>> np.add.reduce(np.ones((2, 2, 2)), axis=(0, 2), initializer=10)
```
To show that it is only counted once per output element, instead of once per axis.

mhvk replied:
I'm still confused: this example is exactly the same as that two lines above.

hameerabbasi replied:
Ahh... My mistake. Fixed.

mhvk reviewed (changes requested):
Some small comments about ordering, and a larger one about `dtype=object` -- if one passes in the `initializer`, shouldn't it just be used?

Also, more importantly, to be sure this works for classes that define `__array_ufunc__`, could you add a test case for those? (in `test_umath`, `test_ufunc_override_methods`)

eric-wieser commented on numpy/add_newdocs.py:
```diff
@@ -5874,7 +5882,20 @@ def luf(lamdaexpr, *args, **kwargs):
+    array([14., 14.])
+    
+    The initializer argument can be used to calculate empty reductions for
+    ufuncs that don't support it as well.
```
Would maybe read better as "allows reductions of empty arrays where they would normally fail" - I don't like the `as well` at the end here.

Could also show how `np.minimum.reduce([])` fails

hameerabbasi commented:
Is this ready to send to the mailing list?

hameerabbasi commented:
@ahaldane That is an excellent point. In general, I'm of the view that `gufunc`s themselves should be made as flexible as possible, to allow as many operations to be made into `gufunc`s as possible. Since it was your idea, I think you should open the issue of allowing arbitrary kwargs.

eric-wieser commented:
Needs rebase

hameerabbasi commented:
Done rebase.

eric-wieser commented on numpy/lib/nanfunctions.py:
```diff
@@ -1190,8 +1229,8 @@ def _nanquantile_1d(arr1d, q, overwrite_input=False, interpolation='linear'):
-    arr1d, overwrite_input = _remove_nan_1d(arr1d,
-        overwrite_input=overwrite_input)
+    arr1d, overwrite_input = _remove_nan_1d(
+        arr1d, overwrite_input=overwrite_input)
```
I agree with this wrapping, but would prefer to see it left to another PR

hameerabbasi replied:
Done.

eric-wieser replied:
It's showing in the diff just fine - you just did it in two places, and only reverted it in one

hameerabbasi replied:
Ahh, my bad. Fixed.

eric-wieser commented on numpy/lib/tests/test_nanfunctions.py:
```diff
@@ -83,6 +83,17 @@ def test_result_values(self):
+        assert_equal(np.nanmax(mat1, initializer=6), 6.0)
+
+        assert_equal(np.nanmin(mat2, initializer=5), 1.0)
+        assert_equal(np.nanmax(mat2, initializer=6), 6.0)
```
Let's not use `mat` to avoid confusion with `np.matrix`. `arr` or `a` is pretty typical in tests

hameerabbasi replied:
Done.

hameerabbasi commented:
Hmm, it seems my last commit isn't registering in the diff for some reason. It fixes your comment about line-wrapping.

hameerabbasi commented:
@eric-wieser Could you somehow re-trigger the diff computation? It seems my last commit wasn't taken into account.

@mhvk Do you still have changes to request? Because I think I met your original questions.

mhvk commented on numpy/core/fromnumeric.py:
```diff
@@ -1905,7 +1913,7 @@ def sum(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
         return res
 
-    return _wrapreduction(a, np.add, 'sum', axis, dtype, out, keepdims=keepdims)
+    return _wrapreduction(a, np.add, 'sum', axis, dtype, out, keepdims=keepdims, initializer=initializer)
```
This line is too long.

mhvk commented on numpy/core/fromnumeric.py:
```diff
@@ -2286,11 +2301,25 @@ def amax(a, axis=None, out=None, keepdims=np._NoValue):
+    5
     """
-    return _wrapreduction(a, np.maximum, 'max', axis, None, out, keepdims=keepdims)
+    return _wrapreduction(a, np.maximum, 'max', axis, None, out, keepdims=keepdims, initializer=initializer)
```
Also too long.

mhvk commented on numpy/core/fromnumeric.py:
```diff
@@ -2234,6 +2242,13 @@ def amax(a, axis=None, out=None, keepdims=np._NoValue):
         exceptions will be raised.
 
+    initializer : scalar, optional
+        The maximum value of an output element. Must be present to allow
```
This actually becomes the minimum value the maximum will have, right? Maybe rephrase to make that clearer (and the same for `amin` below).

...

Must admit I cannot easily think of a better way of phrasing this. But at least change maximum -> minimum here.

hameerabbasi replied:
Actually, this is correct. It's the maximum value that the minimum can hold.

mhvk commented on numpy/core/fromnumeric.py:
```diff
@@ -2374,8 +2409,21 @@ def amin(a, axis=None, out=None, keepdims=np._NoValue):
+    array([-50,   0])
+
+    Notice that the initializer is used as one of the elements for which the
+    maximum is determined, unlike for the default argument Python's max
```
maximum -> minimum, max->min (too much copy&paste...)

mhvk commented on numpy/core/fromnumeric.py:
```diff
@@ -2374,8 +2409,21 @@ def amin(a, axis=None, out=None, keepdims=np._NoValue):
+    6
     """
-    return _wrapreduction(a, np.minimum, 'min', axis, None, out, keepdims=keepdims)
+    return _wrapreduction(a, np.minimum, 'min', axis, None, out, keepdims=keepdims, initializer=initializer)
```
too long

mhvk commented on numpy/core/fromnumeric.py:
```diff
@@ -2508,8 +2560,12 @@ def prod(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
+    10
     """
-    return _wrapreduction(a, np.multiply, 'prod', axis, dtype, out, keepdims=keepdims)
+    return _wrapreduction(a, np.multiply, 'prod', axis, dtype, out, keepdims=keepdims, initializer=initializer)
```
line too long

mhvk commented on numpy/core/src/umath/override.c:
```diff
@@ -123,11 +123,11 @@ normalize_reduce_args(PyUFuncObject *ufunc, PyObject *args,
     npy_intp i;
     PyObject *obj;
-    static char *kwlist[] = {"array", "axis", "dtype", "out", "keepdims"};
+    static char *kwlist[] = {"array", "axis", "dtype", "out", "keepdims", "initializer"};
```
line seems too long.

mhvk commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -2831,14 +2831,14 @@ reduce_loop(NpyIter *iter, char **dataptrs, npy_intp *strides,
 static PyArrayObject *
 PyUFunc_Reduce(PyUFuncObject *ufunc, PyArrayObject *arr, PyArrayObject *out,
-        int naxes, int *axes, PyArray_Descr *odtype, int keepdims)
+        int naxes, int *axes, PyArray_Descr *odtype, int keepdims, PyObject *initializer)
```
line again seems too long

mhvk commented on numpy/lib/nanfunctions.py:
```diff
@@ -25,12 +25,11 @@
     'nanmedian', 'nanpercentile', 'nanvar', 'nanstd', 'nanprod',
     'nancumsum', 'nancumprod'
-    ]
+]
```
Total nitpick: can we either not change this (and not remove the empty line above), or move the `]` to the end of the list, i.e., behind `nancumprod`?

mhvk commented on numpy/lib/nanfunctions.py:
```diff
@@ -219,6 +218,12 @@ def nanmin(a, axis=None, out=None, keepdims=np._NoValue):
 
         .. versionadded:: 1.8.0
+    initializer : scalar, optional
+        The maximum value of an output element. Must be present to allow
```
See comment above.

hameerabbasi replied:
Actually, this one is correct. It's for `nanmin`, so it IS the maximum value.

mhvk replied:
This mostly referred to the initial comment of asking for a rewording. But I then later realized I did not have a good concrete suggestion... Anyway, at a minimum, just fix the wrong one for maximum above.

eric-wieser replied:
> Anyway, at a minimum

Unhelpful choice of words...

mhvk commented on numpy/lib/nanfunctions.py:
```diff
@@ -380,26 +396,31 @@ def nanmax(a, axis=None, out=None, keepdims=np._NoValue):
         res = np.amax(a, axis=axis, out=out, **kwargs)
         if mask is None:
             return res
-
```
Just keep the line (it is there for `min` as well).

mhvk commented on numpy/lib/nanfunctions.py:
```diff
@@ -492,7 +513,7 @@ def nanargmax(a, axis=None):
 
 
-def nansum(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
+def nansum(a, axis=None, dtype=None, out=None, keepdims=np._NoValue, initializer=np._NoValue):
```
line too long

mhvk commented on numpy/lib/nanfunctions.py:
```diff
@@ -579,10 +604,15 @@ def nansum(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
-    return np.sum(a, axis=axis, dtype=dtype, out=out, keepdims=keepdims)
+
+    if initializer is not np._NoValue:
+        initializer, mask = _replace_nan(initializer, 0)
```
Since the mask is not used, maybe `initializer, _  = _replace_nan(initializer, 0)`

Actually, do we want to bother to replace it at all? Why let the user silently pass in nonsense?

hameerabbasi replied:
The reasoning here is; since these are `nan`-skipping reductions, they should skip the `nan` initializer as well. Otherwise, they won't have the effect of being nan-skipping, would they?

mhvk replied:
But why apply that to the initializer? I just don't see any use case for the increased complexity, and I do see a case for giving an error/warning when input is unexpected! In other words, I think most if not all cases where `initializer=np.nan` are going to be erroneous, and this might as well lead to an error.

hameerabbasi replied:
I'm not sure, but I do agree this is worth discussing. Maybe we can defer to someone else for this case?

mhvk replied:
@eric-wieser - what do you think? Does it make sense to replace `nan` in the initlializer with `0` here, or should we just let things fail for such input?

eric-wieser replied:
We should choose such that

```
func(arr_1, initializer=func(arr_2, initializer=x))
```
is always treated the same as

```
func(concatenate((arr_1, arr_2)), initializer=x)
```

I think this is maintained in either case?

hameerabbasi replied:
For @mhvk's suggestion, it isn't maintained for `initializer=nan`

eric-wieser replied:
Do you mean for `x == nan`? Can you show the code that fails that test?

mhvk replied:
If in both cases one gets an exception, that would be fine too!

mhvk commented on numpy/lib/nanfunctions.py:
```diff
@@ -644,7 +678,12 @@ def nanprod(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
-    return np.prod(a, axis=axis, dtype=dtype, out=out, keepdims=keepdims)
+
+    if initializer is not np._NoValue:
+        initializer, mask = _replace_nan(initializer, 1)
```
Same as above - I would remove this altogether.

mhvk reviewed (commented):
@hameerabbasi - looks very good and all my remaining comments are nitpicks, except for the fact that I don't think we should replace `initializer=np.nan` with `0` - don't silently accept nonsense!

eric-wieser commented on numpy/lib/tests/test_nanfunctions.py:
```diff
@@ -83,6 +83,17 @@ def test_result_values(self):
+        assert_equal(np.nanmax(arr1, initializer=6), 6.0)
+
+        assert_equal(np.nanmin(arr2, initializer=5), 1.0)
+        assert_equal(np.nanmax(arr2, initializer=6), 6.0)
```
Can you add some tests of `np.nanmax([], initializer=np.nan)`, `np.nanmax([np.nan], initializer=np.nan)`, and `np.nanmax([np.nan], initializer=0)`?

hameerabbasi replied:
What me and @mhvk differ on is the behavior of `np.nanmax([0], initializer=np.nan)`. I believe this should be zero, he says it should err or warn; and return `nan`.

hameerabbasi replied:
And sure, I can add those tests. Not immediately though. It'd be best if you could settle this debate first.

mhvk replied:
Just to be clear, in my ideal world `initializer=np.nan` just raises an exception. I haven't checked what it actually does if you remove the turn-it-into-0 stanza.

eric-wieser replied:
Perhaps we should leave the nanfunctions to another PR?

hameerabbasi replied:
> Perhaps we should leave the nanfunctions to another PR?

I'm okay with that.

> Just to be clear, in my ideal world initializer=np.nan just raises an exception. I haven't checked what it actually does if you remove the turn-it-into-0 stanza.

It will make everything `nan`.

mhvk replied:
> It will make everything `nan`.

Not my ideal world, but I think preferable to silently changing it. But maybe it is indeed best to just move the `nanfunctions` to another PR, so we can merge this one.

hameerabbasi commented:
Are there still changes needed here or can we move the NaN functions to another PR and merge this one?

mhvk commented:
I'd prefer not touching `nanfunctions` here beyond passing on the `initializer` argument (and even that could be pushed on).

eric-wieser commented:
Better not to add it to the nanfunctions at all yet, so we're not locked
into an API we regret.

Let's take them out for now, and ask the mailing list for approval for
initializer in the regular reductions.

On Sun, Mar 4, 2018, 20:34 Marten van Kerkwijk <notifications@github.com>
wrote:

> I'd prefer not touching nanfunctions here beyond passing on the
> initializer argument (and even that could be pushed on).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/numpy/numpy/pull/10635#issuecomment-370261601>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAZ9LGBRfb3RBbaqd1pXt0A5tATS2CbNks5tbE_RgaJpZM4SMMhf>
> .
>

hameerabbasi commented:
@eric-wieser Should I do that? I don't know git as well as you.

hameerabbasi commented:
I've removed the Nan changes and moved them into a separate branch. I'll open another PR once this one is merged.

hameerabbasi commented:
Any more changes required here? cc @eric-wieser @mhvk

eric-wieser commented on numpy/core/tests/test_ufunc.py:
```diff
@@ -941,6 +946,27 @@ def test_identityless_reduction_noncontig_unaligned(self):
+        assert_equal(np.minimum.reduce([], initializer=np.inf), np.inf)
+        assert_equal(np.maximum.reduce([], initializer=-np.inf), -np.inf)
+
+        # Rendom tests
```
Typo?

hameerabbasi replied:
Fixed!

eric-wieser commented on numpy/core/tests/test_ufunc.py:
```diff
@@ -941,6 +946,27 @@ def test_identityless_reduction_noncontig_unaligned(self):
+
+        # Rendom tests
+        assert_equal(np.minimum.reduce([5], initializer=4), 4)
+        assert_equal(np.maximum.reduce([4], initializer=5), 5)
```
Can you add the other two tests with `minimum` and `maximum` swapped?

hameerabbasi replied:
Done!

hameerabbasi replied:
Added!

eric-wieser commented on numpy/core/tests/test_ufunc.py:
```diff
@@ -941,6 +946,27 @@ def test_identityless_reduction_noncontig_unaligned(self):
+            np.minimum.reduce([])
+
+        with assert_raises(ValueError):
+            np.minimum.reduce([], initializer=None)
```
One-liners like this are often spelt `assert_raises(ValueError, np.minimum.reduce, [], initializer=None)` elsewhere in the test suite. Fine as is though.

mhvk replied:
I must admit I find this way much clearer...

hameerabbasi replied:
I agree with @mvhk here that this way is clearer, but my desire for uniformity kind of overrules that.

mhvk commented on numpy/add_newdocs.py:
```diff
@@ -5843,6 +5843,14 @@ def luf(lamdaexpr, *args, **kwargs):
+    initializer : scalar, optional
+        The value with which to start the reduction.
+        Defaults to the `~numpy.ufunc.identity` of the ufunc.
+        If ``None`` is given, then the first element of the reduction is used,
```
This text is confusing for the reader: we state that the default is the identity, but in the top line it notes `initializer=None` and then here it says that for that case the first element is used. Maybe all it needs is "if explicitly given as ``None``"

eric-wieser replied:
"if explicitly given as `None`" is not accurate, since it also implies when `identity is None` and the argument is ommited.

Perhaps _"When `None`, either explicitly or when omitted for a ufunc with no identity, ..."_

hameerabbasi replied:
I have a feeling that the complications in the docs are getting out of hand, and may make them hard to parse. Perhaps keep as-is, since we already mention that it defaults to the identity? And change the signature to say `ufunc.identity` instead of `None`?

mhvk replied:
@hameerabbasi - I was thinking the same thing: the problem is really the signature. It makes sense to change it to `ufunc.identity` to me.

hameerabbasi replied:
Done!

mhvk commented on numpy/core/tests/test_ufunc.py:
```diff
@@ -493,6 +493,11 @@ def test_sum_complex(self):
 
+    def test_sum_initializer(self):
+        assert_equal(np.sum([3], initializer=2), 5)
+        assert_almost_equal(np.sum(np.ones((2, 3, 5)), axis=(0, 2), initializer=2),
```
Why `almost_equal` instead of `equal`? This should be exact also for float.

hameerabbasi replied:
I tend to keep all floating-point stuff `almost_equal`. It's just a habit.

mhvk replied:
It's a nitpick, but I'd like this to be `equal` - it should be exact so we might as well test it as such.

mhvk commented:
Only tiny comments left.

hameerabbasi commented:
I guess I'll send this to the mailing list this evening.

hameerabbasi commented:
Submitted, if you want to participate in the discussion. [Link to main thread](https://mail.python.org/pipermail/numpy-discussion/2018-March/077842.html).

hameerabbasi commented:
@charris This may be a bit late, but I've edited the original comment to explain all the changes I've made.

eric-wieser reviewed (approved):
All looks good to me. Let's leave it on the mailing list for a week, but if there are no opinions, I'm still +1 on merging this. It gives an easy solution to the problem addressed by #10688 too

hameerabbasi commented:
@mhvk Are the changes good? If so, can you approve them? The red is annoying. :P
@eric-wieser Since this is a change in the core Numpy machinery, I feel opinions are important. I'm not against merging, but I don't want it to get reverted down the road for something we didn't think of. :-)

But you're a core dev, so I think you would know.

mhvk commented:
@hameerabbasi - it all looks good! Only one nitpick that would be nice to address. While you are at it, could you squash the commits into one?

hameerabbasi commented:
My OCDs didn't agree with me using `assert_equal` with floating point operations like `sum` which may be inexact... So I converted it to `int64`, and added another test for `assert_almost_equal` which is, in fact inexact.

mhvk commented:
> My OCDs didn't agree with me using `assert_equal` with floating point operations like `sum` which may be inexact... So I converted it to `int64`, and added another test for `assert_almost_equal` which is, in fact inexact.

:smile: - definitely all OK now.

eric-wieser commented:
Perhaps we should revisit the argument name:

* `functools.reduce` actually calls it `initial`, not `initializer`
* `initializer` hints of a function, not a value, to me
* However, `functools.reduce` also takes no kwargs, so what it calls the argument is somewhat moot

@shoyer suggests `initial_val`, which I think I like more than `initializer`, but less than `initial`.

Is there any name we can use with existing kwarg usage?

hameerabbasi commented:
I'm +1 on `initial`. If everyone agrees, I'll make the changes.

mhvk commented:
:+1: for doing the same as `functools.reduce` - even if that is `initializer`! (Indeed, I'd rather change `identity` of the ufuncs to be `initializer`...)

seberg commented:
The offline help calls it `intial` the online `intitializer` and the function seems to not allow kwargs. As a point for `initial`, to me `initializer` sounds like it would be a function and not a value (my point would be, I find it likely the python guys did not think about it too much, since it is just documentation).

eric-wieser commented:
> Indeed, I'd rather change identity of the ufuncs to be initializer...

I don't agree with this - `identity` refers to the special value which is a no-op when passed as an initializer, so needs a different name.

shoyer commented:
Indeed, `initializer` sounds like a function name. For example, that's how TensorFlow uses arguments named `initializer`.

My preference would be for `initial_value`. It's slightly longer than `initial`, but adjectives do not make for good argument names.

hameerabbasi commented:
How about a shortened version: `initval`?

eric-wieser commented:
Not a fan of `initval`.

Myargument against `initial_value` is that `value`, while indeed changing it from an adjective to a noun, is a noun that contains no information - we don't use `out_value`, `x_value`, or `where_value`.

My preference order of the suggestions so far, from best to worst, is:

* `initial`
* `initial_value`
* `initial_val`
* `initializer`
* `initval`

mhvk commented:
For me, either `initial` or `initializer` work -- and I think it should be either of those two for consistency with `functools.reduce`.

hameerabbasi commented:
It seems `inital` is the most accepted `kwarg`. If everyone is okay with it shall I change to `initial`?

mhvk commented:
@hameerabbasi - I think it would be best if we had `initial` for `object` dtype as well, and if it is easy to do, might as well do it here. But fine also to push off to another PR.

seberg commented:
Wait, if you push it off to later, you have to make it an *error* now, that might be an option (if not too nice).

hameerabbasi commented:
I agree with @seberg, I'd prefer to do it here. If someone can point me to the right bit of code (cc @eric-wieser) I'm perfectly willing to do it now.

hameerabbasi commented:
Never mind. My bad, must have been blind. The code was right there. I fixed the object dtype. :-)

mhvk commented on numpy/add_newdocs.py:
```diff
@@ -5884,7 +5891,24 @@ def luf(lamdaexpr, *args, **kwargs):
            [ 9, 13]])
-
+           
+    You can use the initial argument to initialize the reduction with a
```
Maybe put double-backquote around `initial` - otherwise "initial argument" reads as if the first argument is meant.

Or avoid it altogether with:
```
You can also start the reduction with a different initial value.
```

seberg commented:
Can you please make sure that you have tests testing high dimensional arrays (with axis keyword) on both (fast and slow in memory) axes?

hameerabbasi commented:
@seberg This was discussed above. I tested a 3D array with `axis=(0, 2)` already, and it works fine.

mhvk commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -3061,32 +3063,44 @@ PyUFunc_Reduce(PyUFuncObject *ufunc, PyArrayObject *arr, PyArrayObject *out,
-        Py_DECREF(identity);
-        identity = Py_None;
-        Py_INCREF(identity);
+    if (!initial_given && initial != Py_None && PyArray_ISOBJECT(arr) && PyArray_SIZE(arr) != 0) {
```
I would move this if statement inside the `if (initial == NULL)` clause above (and then test `identity`), so that you can do without the `initial_given` variable.

mhvk reviewed (commented):
Looks good, but one small comment on the docstring, and a somewhat larger one on the code. Also, does need test cases for `dtype=object`...

mhvk commented:
@hameerabbasi - OK, those changes look good! I think this all looks OK now.

hameerabbasi commented on numpy/add_newdocs.py:
```diff
@@ -5853,6 +5853,13 @@ def luf(lamdaexpr, *args, **kwargs):
         .. versionadded:: 1.7.0
+    initial : scalar, optional
+        The value with which to start the reduction.
+        Defaults to the `~numpy.ufunc.identity` of the ufunc.
```
Nice catch! Yes, it's `None` for object arrays.

hameerabbasi replied:
If we want to get pedantic we also have to separate the case when the array is empty and `dtype=object`, in which case `ufunc.identity` is used.

seberg replied:
Yes, we want to be pendantic. Also be pendantic about `None`? `None` could be a valid "identity" parameter, we should think about checking if the parameter was passed at all (even if it is a bit strange maybe), or use another "default" like NoValue?

hameerabbasi replied:
I really can't see a use case for making `None` a valid initial value. We have a way of checking if something is passed.

But your comment does remind me of something else. In `numpy/core/_methods.py`, we're passing `initial` explicitly, which may result in issues with `dtype=object`.

@eric-wieser Any insights into how to solve that? The only one I can see is adopting @seberg's suggestion of defaulting to some sort of no-value.

eric-wieser replied:
This was the can of worms I didn't want to open in this PR, but leave till later.

One solution would be to introduce a `np.ufunc.NoIdentity` sentinel, which replaces the current use of None.

hameerabbasi replied:
I agree that we should delay it, but also that we need to fix the `_methods` thing. Or I can just revert the recent commits.

hameerabbasi replied:
I can introduce a few more lines in there, default it to `np._NoValue`, but that'd make it slow. ☹️

hameerabbasi replied:
Or I can just move `None` to be a no-op for now and leave all these issues till later. That seems like the best way forward to me at this point.

eric-wieser replied:
> I really can't see a use case for making None a valid initial value

I can - `np.logical_or.reduce(obj_array, initial=None)` picks the first truthy value from a sequence, else returns None.

hameerabbasi commented:
Since supporting `dtype=object` is a can of worms (as @eric-wieser so aptly put it), I'd like to revert the object dtype stuff. This will effectively make `initial` equivalent to `default` for `dtype=object`. Or we can follow @seberg's suggestion and make it an error until we can properly support it.

@eric-wieser What would be the best way to make it an error within C code with a nice description?

mhvk commented:
I think what you, the behaviour for objects is close enough to what the final state should be than what would be the case without those extra commits. So, I would suggest to merge and do a follow-up PR on the details.

hameerabbasi commented:
+1 for finally merging.

seberg commented:
Well, I don't know and I do not want to think about it. Close enough for me is close enough, if assuming that someone will use the current behaviour, we can still get to a reasonable final state without problems at some point. I hate adding the next kludge that future us gets annoyed about.

Right now I am a bit confused what you are suggesting as final state for the object support....

seberg commented:
That said, None as the "Error out" value, is maybe good even for object. Not ideal in some sense but the other solutions seem not much nicer. Plus, I am not sure it would even manoeuvre into a tricky to resolve issue for the future.

eric-wieser commented:
I think the cause of our problems is how the `.identity` is defined, and we'd do better to free ourselves from that constraint. What I'd propose then is:

* Use `.reduce(initial=np._NoValue)` to indicate "no identity"
* Make `.reduce(obj_array, initial=some_value)` use the "initial" logic, not the "default" logic
* Allow `None` to be passed like any other value
* Remove the `initial=ufunc.identity` documentation, and just remark in the documentation of `initial` that _"If the ufunc has no identity or the dtype is object, this default to None - otherwise it defaults to ufunc.identity"_

hameerabbasi commented:
@eric-wieser Would you happen to know how I'd access `np._NoValue` inside the C code for this check? Doesn't seem terribly hard to do.

eric-wieser commented on numpy/core/src/umath/override.c:
```diff
@@ -152,6 +155,9 @@ normalize_reduce_args(PyUFuncObject *ufunc, PyObject *args,
             obj = PyTuple_GetSlice(args, 3, 4);
         }
+        if (i == 5 && obj == NoValue) {
+            continue;
```
A comment saying this is the `initial` argument would be good.

hameerabbasi replied:
Sure. Any idea why the Windows build is failing?

eric-wieser commented:
Last two commit messages are very wrong!

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -3036,6 +3036,8 @@ PyUFunc_Reduce(PyUFuncObject *ufunc, PyArrayObject *arr, PyArrayObject *out,
     /* These parameters come from a TLS global */
     int buffersize = 0, errormask = 0;
+    static PyObject *NoValue = NULL;
+    npy_cache_import("numpy", "_NoValue", &NoValue);
```
Ah, this should actually be `if (... < 0) { return NULL; }` (wrapped as normal), in case the import fails

hameerabbasi replied:
`npy_cache_import` has a return type of `void`, so we can't do that.

eric-wieser replied:
Ah, in that case, I mean you should check that `NoValue != NULL`

eric-wieser commented on numpy/core/src/umath/override.c:
```diff
@@ -137,6 +137,9 @@ normalize_reduce_args(PyUFuncObject *ufunc, PyObject *args,
         return -1;
     }
 
+    static PyObject *NoValue = NULL;
```
C90 requires this be declared at the top of the function with the other variables (which is why the test fails)

hameerabbasi commented:
I think this now has the desired behavior for all cases. The `np.NoIdentity` thing is a matter for a separate PR, this PR keeps the convention that `None` means "no initial value".

seberg commented:
Won't look through the code, at least not now. It seems it still needs release notes I guess? So I understand correctly it is now:
 * Initial value for the reduction for all (almost) cases, including object
 * `None` will cause an error (even for object)
 * default is `NoValue` and will cause old behaviour.

Seems good to me as a change.

hameerabbasi commented:
That's correct:

- `initial` behaves as an initial value in all cases now, even for `dtype=object`. Previously, it would behave as the default value for `dtype=object` and initial value otherwise.
- `None` means "no initial value, err if empty, otherwise use the start of the array". 
- `np._NoValue` causes the old behavior.

I'll add release notes now.

hameerabbasi commented:
Rebased and tests added for `dtype=object`. Tell me when merging is imminent so I can squash all the commits.

mhvk commented:
@hameerabbasi - I'd gladly merge this (maybe after one more day for last comments), so do go ahead and squash!

hameerabbasi committed:
ENH: Implement initial kwarg for ufunc.add.reduce

hameerabbasi commented:
@mhvk Ping. 😀

mhvk commented:
Merged. Thanks, @hameerabbasi!

hameerabbasi commented:
Thanks so much to @eric-wieser and @mhvk especially for babying me through my first big PR. I owe you!
